services/src/main/java/org/keycloak/credential/PasswordCredentialProvider.java

```diff
@@ -208,6 +197,26 @@ public boolean isValid(RealmModel realm, UserModel user, CredentialInput input)
         return true;
     }
 
+    private void rehashPasswordIfRequired(KeycloakSession session, RealmModel realm, UserModel user, CredentialInput input, PasswordCredentialModel password) {
+        PasswordPolicy passwordPolicy = realm.getPasswordPolicy();
+        PasswordHashProvider provider;
+        if (passwordPolicy != null && passwordPolicy.getHashAlgorithm() != null) {
```

Comment:
nitpick: looks like this logic around "give this specific provider or else default" is used in different places. Perhaps, we should have it within the `PasswordPolicy`?

Reply:
I agree, but couldn't really do that in a nice way without introducing something like session.hash()...

Leaving it for now, but will see if I can come up with something as a follow-up